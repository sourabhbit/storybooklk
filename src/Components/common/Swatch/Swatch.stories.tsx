import { ComponentMeta, ComponentStory } from '@storybook/react'
import { generateArgTypes } from 'Helpers/Storybook'
import Swatch from './index'

export default {
	title: 'Components/common/Swatch',
	component: Swatch,
	argTypes: {
		id: generateArgTypes({
			type: 'string',
			required: true,
			description: 'String used as unique identifier',
			category: 'General',
		}),
		colors: generateArgTypes({
			type: 'array',
			required: true,
			description: `color options array, eg: ['red', 'blue', '#c4c4c4']`,
			category: 'Component',
		}),
		onmouseOver: generateArgTypes({
			type: 'function',
			required: false,
			description: 'onmousehover handler for swatch',
			category: 'Events',
		}),
	},
} as ComponentMeta<typeof Swatch>

const Template: ComponentStory<typeof Swatch> = (args) => <Swatch {...args} />

export const Overview = Template.bind({})
Overview.args = {
	id: 'swatch',
	colors: [],
}
