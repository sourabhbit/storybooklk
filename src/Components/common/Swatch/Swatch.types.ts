export interface SwatchTypes {
	id: string
	colors: string[]
	onmouseOver?: () => void
}
