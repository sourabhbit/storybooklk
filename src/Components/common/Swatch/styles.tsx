import styled from 'styled-components'

export const SwatchWrapper = styled.div`
	display: flex;
	flex-direction: row;
	overflow-x: auto;
	white-space: nowrap;
	gap: 5px;
`
export const SwatchItemWrapper = styled.div``
export const SwatchItem = styled.div<{
	color: string
	selected: boolean
}>`
	min-height: 12px;
	min-width: 12px;
	/* margin: 0 8px; */
	border-radius: 50%;
	background-color: ${(props) => props.color};
	pointer-events: auto;
	cursor: pointer;
	box-sizing: border-box;
	border: ${(props) =>
		props.selected ? '1px solid var(--dark-blue-100)' : 'none'};
	box-shadow: ${(props) =>
		props.selected ? 'inset 0px 0px 0px 2px var(--warm-grey-25)' : 'none'};
`
