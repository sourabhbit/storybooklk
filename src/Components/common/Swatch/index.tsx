import { SwatchTypes } from './Swatch.types'
import { SwatchWrapper, SwatchItem } from './styles'
import { useEffect, useState } from 'react'
import { action } from '@storybook/addon-actions'

const Swatch = ({ id, colors }: SwatchTypes) => {
	const [selected, setSelected] = useState<boolean[]>([])
	useEffect(() => {
		if (colors !== undefined && colors.length >= 1) {
			setSelected((prev) => [true, ...prev.slice(1)])
			action('default')({
				colorCode: colors[0],
			})
		}
	}, [colors])
	const handleHover = (
		e: React.MouseEvent<HTMLDivElement, MouseEvent>,
		idx: number,
		item: string
	) => {
		const newArr = selected.map(() => false)
		newArr[idx] = true
		setSelected(newArr)
		action('color changed')(e, { colorCode: item })
	}

	if (colors !== undefined && colors.length > 0) {
		return (
			<SwatchWrapper data-testid="swatch-test" id={id}>
				{colors.map((item, idx) => (
					<SwatchItem
						data-testid="swatch-item"
						color={item}
						key={idx}
						selected={selected[idx]}
						onMouseOver={(e) => handleHover(e, idx, item)}
					/>
				))}
			</SwatchWrapper>
		)
	} else {
		return null
	}
}

export default Swatch
