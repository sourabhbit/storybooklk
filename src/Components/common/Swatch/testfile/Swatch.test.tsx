import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import Swatch from '../index'
import { addons, mockChannel } from '@storybook/addons'

addons.setChannel(mockChannel())
const customColors = [
	'#705C65',
	'#506E93',
	'#FFE091',
	'#EEEBEB',
	'#C4C4C4',
	'#4B5944',
	'#908942',
]
const props = {
	id: 'swatch',
	onmouseover: jest.fn(),
	colors: customColors,
}

describe('<Swatch />', () => {
	let _asFragment: any = null
	let _rerender: any = null

	beforeEach(() => {
		const { asFragment, rerender } = render(<Swatch {...props} />)
		_asFragment = asFragment
		_rerender = rerender
	})

	afterAll(() => {
		_asFragment = null
	})

	it('Should render component', () => {
		const container = screen.getByTestId('swatch-test')
		expect(container).toBeInTheDocument()
	})

	it('component should not render if no colors', () => {
		_rerender(<Swatch {...props} colors={[]} />)
		const container = screen.queryByTestId('swatch-test')
		expect(container).not.toBeInTheDocument()
	})

	it('Should take a snapshot', () => {
		expect(_asFragment()).toMatchSnapshot()
	})
})
