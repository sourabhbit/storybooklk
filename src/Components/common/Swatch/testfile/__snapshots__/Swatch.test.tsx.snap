// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Swatch /> Should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL cLTMZH"
    data-testid="swatch-test"
    id="swatch"
  >
    <div
      class="sc-dkPtRN krUQxC"
      color="#705C65"
      data-testid="swatch-item"
    />
    <div
      class="sc-dkPtRN kCFcDT"
      color="#506E93"
      data-testid="swatch-item"
    />
    <div
      class="sc-dkPtRN bTZgJu"
      color="#FFE091"
      data-testid="swatch-item"
    />
    <div
      class="sc-dkPtRN eFLBJH"
      color="#EEEBEB"
      data-testid="swatch-item"
    />
    <div
      class="sc-dkPtRN loNMqA"
      color="#C4C4C4"
      data-testid="swatch-item"
    />
    <div
      class="sc-dkPtRN lafMCh"
      color="#4B5944"
      data-testid="swatch-item"
    />
    <div
      class="sc-dkPtRN hktMER"
      color="#908942"
      data-testid="swatch-item"
    />
  </div>
</DocumentFragment>
`;
